vsphere/resource_vsphere_license.go

```diff
@@ -0,0 +1,246 @@
+			"license_key": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+			"labels": &schema.Schema{
+				Type:     schema.TypeMap,
+				Optional: true,
+			},
+
+			// computed properties returned by the API
+			"edition_key": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+			"total": &schema.Schema{
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+			"used": &schema.Schema{
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceVSphereLicenseCreate(d *schema.ResourceData, meta interface{}) error {
+	log.Println("[INFO] Running the create method")
+
+	client := meta.(*govmomi.Client)
+	manager := license.NewManager(client.Client)
+
+	key := d.Get("license_key").(string)
+
+	log.Println(" [INFO] Reading the key from the resource data")
+	finalLabels := make(map[string]string)
+
+	if labels, ok := d.GetOk("labels"); ok {
+		labelMap := labels.(map[string]interface{})
+		for key, value := range labelMap {
+			finalLabels[key] = value.(string)
+		}
+	}
+
+	var info types.LicenseManagerLicenseInfo
+	var err error
+	switch t := client.ServiceContent.About.ApiType; t {
+	case "HostAgent":
+		// Labels are not allowed in ESXi
+		if len(finalLabels) != 0 {
+			return errors.New("Labels are not allowed in ESXi")
+		}
+
+		info, err = manager.Update(context.TODO(), key, finalLabels)
+	case "VirtualCenter":
+		info, err = manager.Add(context.TODO(), key, finalLabels)
+	default:
+		return fmt.Errorf("unsupported ApiType: %s", t)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	if err = DecodeError(info); err != nil {
+		return err
+	}
+
+	// This can be used in the read method to set the computed parameters
+	d.SetId(info.LicenseKey)
+
+	return resourceVSphereLicenseRead(d, meta)
+}
+
+func resourceVSphereLicenseRead(d *schema.ResourceData, meta interface{}) error {
+	log.Println("[INFO] Running the read method")
+
+	client := meta.(*govmomi.Client)
+	manager := license.NewManager(client.Client)
+
+	if info := getLicenseInfoFromKey(d.Get("license_key").(string), manager); info != nil {
+		log.Println("[INFO] Setting the values")
+		d.Set("edition_key", info.EditionKey)
+		d.Set("total", info.Total)
+		d.Set("used", info.Used)
+		d.Set("name", info.Name)
+		d.Set("labels", keyValuesToMap(info.Labels))
+
+	} else {
+		return ErrNoSuchKeyFound
+	}
+
+	return nil
+}
+
+// resourceVSphereLicenseUpdate check for change in labels of the key and updates them.
+func resourceVSphereLicenseUpdate(d *schema.ResourceData, meta interface{}) error {
+	log.Println("[INFO] Running the update method")
+
+	client := meta.(*govmomi.Client)
+	manager := license.NewManager(client.Client)
+
+	if key, ok := d.GetOk("license_key"); ok {
+		licenseKey := key.(string)
+		if !isKeyPresent(licenseKey, manager) {
+			return ErrNoSuchKeyFound
+		}
+
+		if d.HasChange("labels") {
```

Comment:
Again, this is the logic that should probably be spun off into its own function so it can be used in `Create` as well to add the labels as a separate step after the initial addition of the license is performed.